### PR TITLE
cafs-cached ro FUSE fs

### PR DESCRIPTION
**Pull request by ransomw1c**
address #158

### CODE REVIEW

**Comment by ransomw1c** [comment]
this is still very early progress.  opening the merge request now for communication purposes.

**Comment by ransomw1c** [comment]
i still consider this very much work-in-progress with some amount of cleanup at all layers (examples: tee writes to caching cafs, design updates to creating rofs instances, testing of CLI .. and everywhere).

yet conceptually, this is ready for a test drive with
```
umount /tmp/mmp ; rm -r /tmp/mmfs /tmp/mmp ; mkdir /tmp/mmfs ; mkdir /tmp/mmp ; ./cmd/datamon/run_datamon bundle mount --repo ransom-datamon-test-repo --destination /tmp/mmfs --mount /tmp/mmp --cache
```
then observe that as files are read at `/tmp/mmp` the cache is populated at `/tmp/mmfs`.

**Comment by ransomw1c** [comment]
similarly, the incantation for tests is

```
go build && umount /tmp/mmp; rm -r /tmp/mmp /tmp/mmfs ; go test -v -tags fuse_cli -run 'TestBundleMountCached$'
```

File: pkg/core/fs.go [hunk 1]
```diff
@@ -50,7 +52,12 @@ type MutableFS struct {
 }
 
 // NewReadOnlyFS creates a new instance of the datamon filesystem.
+// todo: control flow regarding whether to cache pivots on this function.  ideas:
```
Comment (ransomw1c) [inline pkg/core/fs.go]: this api design quandry is the only additional `todo` incurred with this diff .. would appreciate opinions regarding this code design decision in particular

**Comment by ransomw1c** [comment]
particular questions upon letting this marinate and then revisiting:

* use `mv` for atomicity of (linux) fs operations?
*

File: pkg/cafs/cafs.go [hunk 2]
```diff
@@ -252,3 +259,124 @@ func IsRootKey(fs storage.Store, key Key, leafSize uint32) bool {
...
+	}
+}
+
+type cachedFs struct {
+	fsImpl           Fs
+	cacheStore       storage.Store
+	cafsKey2StoreKey func(Key) (string, error)
+}
+
+func (fs *cachedFs) Put(ctx context.Context, src io.Reader) (int64, Key, []byte, bool, error) {
```
Comment (kerneltime) [inline pkg/cafs/cafs.go]: Do we need to implement Put? Also, we should not be using ioutil.ReadAll interface.

File: pkg/cafs/cafs.go [hunk 3]
```diff
@@ -252,3 +259,124 @@ func IsRootKey(fs storage.Store, key Key, leafSize uint32) bool {
...
+		return 0, Key{}, nil, false, err
+	}
+	err = fs.cacheStore.Put(ctx, storeKey, &bytesBuf, storage.IfNotPresent)
+	if err != nil {
+		return 0, Key{}, nil, false, err
+	}
+	return written, key, keys, found, err
+}
+
+// ??? thread-safety
```
Comment (kerneltime) [inline pkg/cafs/cafs.go]: For thread safety, the cacheStore can write to a file and move into place. These are CAFS files so safe in terms of over writes.

File: pkg/cafs/cafs.go (hunk 2 above)
Comment (ransomw1c) [reply pkg/cafs/cafs.go]: as of 17e09219ea37 , this falls back to the decorated `fsImpl`.

File: pkg/core/fs.go (hunk 1 above)
Comment (kerneltime) [reply pkg/core/fs.go]: The property for what kind of bundle it is makes sense. Is it pre fetched, or streamed.

File: pkg/cafs/cafs.go (hunk 3 above)
Comment (ransomw1c) [reply pkg/cafs/cafs.go]: i've added this to #158 and will open a separate request to this effect

File: pkg/core/fs.go (hunk 1 above)
Comment (ransomw1c) [reply pkg/core/fs.go]: deferring this for now..

File: pkg/cafs/cafs.go (hunk 3 above)
Comment (ransomw1c) [reply pkg/cafs/cafs.go]: i made a `storage.Store` implementation that relies on move of the underlying `afero.Fs` to provide atomicity and used it in `bundle_mount.go` as suggested here.

considering move as the atomic part of filesystem operations considered insightful.

File: pkg/core/fs.go (hunk 1 above)
Comment (ransomw1c) [reply pkg/core/fs.go]: i've gone with the option of consolidating the `NewReadOnlyFS` functions according to `BundleOption`s

**Pull request closed without merging**